test: add Product filter and property test

vanpham-sw opened the pull request:
test: add ProductFilter

<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfill our contribution guideline (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?


### 2. What does this change do, exactly?


### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).
<!-- Examples:
- closes #123  - closes the issue #123 when the PR is merged
- relates #123 - relates to the issue #123

In case of issue existing only on Jira, link to the Jira issue.
- Jira issue: https://shopware.atlassian.net/browse/NEXT-123
-->

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

vanpham-sw committed:
test: add Product filter and property test

test: add ProductFilter

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

frobel commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+        await ShopAdmin.goesTo(AdminProductDetail.url(variantProductColor.at(2).id));
+        await ShopAdmin.page.waitForLoadState('domcontentloaded');
+        await ShopAdmin.goesTo(AdminProductDetail.url(variantProductSize.at(2).id));
+        await ShopAdmin.page.waitForLoadState('domcontentloaded');
```
Is that step really necessary because to create an AdminPage Context tooks lot of time and I see here in that lines no benefit.

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

frobel commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+        await manufacturerLocator.click();
+        await ShopCustomer.expects(manufacturerLocator).toBeChecked();
+        await ShopCustomer.expects(StorefrontHome.productItemNames).toHaveCount(1);
+        await ShopCustomer.expects(StorefrontHome.productItemNames).toHaveText(
```
Do you test here if any color is available? Would it be better to check directly if the product contains red because you expecting red (because of the filter)

ocavli commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+    const propertyGroupsColor: PropertyGroup[] = [];
+    const propertyGroupsText: PropertyGroup[] = [];
+    propertyGroupsColor.push(color);
+    propertyGroupsText.push(size);
```
Instead of pushing elements to an array after declaration, is it possible to directly initializing the array with values?
 
` const propertyGroupsColor = [color];`

frobel commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+        await ShopCustomer.expects(StorefrontHome.manufacturerFilter).toBeEnabled();
+        await ShopCustomer.expects(StorefrontHome.freeShippingFilter).toBeEnabled();
+        await ShopCustomer.expects(StorefrontHome.priceFilterButton).toBeEnabled();
+        await ShopCustomer.expects(StorefrontHome.productItemNames).toHaveCount(3);
```
That step would lead into an error if another product is available which was not generated by you. (So I had it locally, that another products where available)

frobel commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+
+    await test.step('Verify property display on the product detail page', async () => {
+        await ShopCustomer.goesTo(StorefrontHome.url());
+        await StorefrontHome.productItemNames.nth(0).click();
```
What if another product is on first place..then we have a problem

sguder replied:
+1

frobel commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+    });
+});
+
+test('Customer should be able to create a new property display it on the product detail page', async ({
```
Do you think we should put it in another file because the file is named ProducFilter, but you test here Variant options visibility :thinking:

sguder replied:
I was confident we already had a test that covers this, but turned out we don't.. But yes, I would also put it into a separate file

sguder commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+        await ShopCustomer.expects(manufacturerLocator).toBeChecked();
+        await ShopCustomer.expects(StorefrontHome.productItemNames).toHaveCount(1);
+        await ShopCustomer.expects(StorefrontHome.productItemNames).toHaveText(
+            new RegExp(variantProductColor.map((product) => product.name).join('|'))
```
Don't know if it has to be done this way, but using RegExpdoesn't help to make the test readable. Could this be done another way to help readers with less programming background to understand what happens here?

sguder replied:
Same to the later occurences of RegExp ofc

ocavli commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,219 @@
+    await test.step('Select another manufacturer and verify that a different filter is disabled', async () => {
+        await StorefrontHome.manufacturerFilter.click();
+        const manufacturerLocator = await StorefrontHome.getFilterItemByFilterName(sizeManufacturer.name);
+        await manufacturerLocator.click();
```
Could it be possible to create a task, like selectFilterOption, to avoid writing similar steps?

`selectFilterOption(filterButton: Locator, optionName: string) {}` sth like that...

vanpham-sw committed:
Update ProductFilter.spec.ts

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

vanpham-sw committed:
working on comment

test on saas

Update package-lock.json

working on comment

working on PR's comment

working on comment

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

ocavli commented on tests/acceptance/package-lock.json:
```diff
@@ -671,9 +671,9 @@
-      "integrity": "sha512-IbXw/1I8mnBfMrLGerDMBQr+tpOWvcEZl9hf8tUIdEDmQZRYYgVcQ7mizaZnNH4CYaosLdlwrf3CTzagZts0rg==",
+      "version": "11.16.2",
+      "resolved": "https://registry.npmjs.org/@shopware-ag/acceptance-test-suite/-/acceptance-test-suite-11.16.2.tgz",
+      "integrity": "sha512-jNYHPns0dBMwOL7viKVL8dBdwgnDbtayfjx1HHWA9N33dmBL8T4hEvNmTX7n7qJEh7tN0Ds8/glFs+Ll8BfpPA==",
```
This one should already be updated with 11.17.1

vanpham-sw committed:
Merge branch 'trunk' into next-40303/test-product-filter

sguder commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,202 @@
+    });
+});
+
+test('Should filter by rating and disable filter options those cannot combine with rating', async ({
```
Having problems understanding what the test is doing, a complete sentence would be nice here

sguder commented on tests/acceptance/tests/Product/ProductFilter.spec.ts:
```diff
@@ -0,0 +1,202 @@
+        await ShopCustomer.expects(await StorefrontHome.getFilterButtonByFilterName(color.name)).toBeVisible();
+    });
+
+    await test.step('Select a rating and verify that unavailable filter is disabled and products are filtered', async () => {
```
Same here, complete sentences

vanpham-sw committed:
update test description

vanpham-sw committed:
Merge branch 'next-40303/test-product-filter' of https://github.com/shopware/shopware into next-40303/test-product-filter
